### PR TITLE
Add a FlashFile board interface to manipulate board flash dumps

### PR DESCRIPTION
**Pull request by lschuermann**
Adds a new board interface used to manipulate board flash dumps. This can be useful for building an image containing a kernel and apps for flashing with other tools, for instance with proprietary toolchains or booting an image via TFTP. Furthermore, this can be useful for building images which are embedded into FPGA bitstreams or production boards.

In addtion to that, this generalizes the address translation mechanisms introduced for OpenOCD to easily work with all boards. `BoardInterface` is extended to have an attribute `address_translator` which can hold a function if address translation is required. A helper method `translate_address` can then be used in the `BoardInterface` implementations to perform the translation, if necessary.

### CODE REVIEW

**Comment by bradjc** [comment]
I added some updates.

Can you confirm that adding `"litex_arty": {"start_address": 0x41000000},` to tockloader.py is correct?

Also, can we update libtock-c so that valid tbfs for the board are created?

File: tockloader/board_interface.py [hunk 1]
```diff
@@ -127,6 +129,31 @@ class BoardInterface:
...
+            "address_translator": lambda addr: addr - 0x40000000,
+            "no_attribute_table": True,
+            "flash_file": {
+                # Set to the maximum RAM size, as the LiteX bootloader will
+                # update the flash image into RAM.
+                "max_size": 0x10000000,
+            },
+        },
+        "litex_sim": {
+            "description": "LiteX SoC running on an Arty-A7 board",
```
Comment (bradjc) [inline tockloader/board_interface.py:148]: Simulated?

Comment (lschuermann) [reply tockloader/board_interface.py:148]: Whoops, sorry.

**Comment by lschuermann** [comment]
> I added some updates.

Thanks!

> Can you confirm that adding `"litex_arty": {"start_address": 0x41000000},` to tockloader.py is correct?

Yes, that looks correct. While we're at it, I have taken the liberty to further add the Verilated simulation. This also eases testing these changes.

> Also, can we update libtock-c so that valid tbfs for the board are created?

We can add a corresponding entry to `TOCK_TARGETS`, yes. However, I was a bit wary of that given that before this -- with my very hacky way of getting RISC-V apps to work -- I frequently needed to change the RAM address after kernel rebuilds. Just to make sure, to add these targets I would just leave a bit of space after `sappmem` so that we aren't risking to collide with the potentially growing kernel memory region?

**Comment by bradjc** [comment]
> Just to make sure, to add these targets I would just leave a bit of space after `sappmem` so that we aren't risking to collide with the potentially growing kernel memory region?

The gap is designed to accommodate a variable sized TBF header. But it would probably also work if the start of app memory shifted? The trick is informing tockloader that the start of apps has changed.

**Comment by lschuermann** [comment]
> The gap is designed to accommodate a variable sized TBF header. But it would probably also work if the start of app memory shifted? The trick is informing tockloader that the start of apps has changed.

I'm afraid I don't quite understand. I'm only talking about RAM, flash is of course at a fixed start address. As far as I know, in case of non-relocatable RISC-V apps, `tockloader` is only concerned about the start of application RAM when it comes to making sure these regions don't overlap with each other -- `tockloader` does not know about `_sappmem`, as that can change with each Tock kernel modification.

What I mean is that between the kernel's `_sappmem` and the start of RAM for the first app, I'd leave a bit of space -- something like 4kB or even 16kB if we can afford it. This way, the kernel's RAM usage can grow without us having to change the `TOCK_TARGETS` config.

As an example, for the LiteX simulation: the kernel has the following application memory region constraints:
```
  9284: 50000000     0 NOTYPE  GLOBAL DEFAULT  ABS _eappmem
  9285: 00100000     0 NOTYPE  GLOBAL DEFAULT    4 _eapps
  9290: 40004000     0 NOTYPE  GLOBAL DEFAULT    6 _sappmem
  9291: 00080000     0 NOTYPE  GLOBAL DEFAULT    4 _sapps
  ```

I would now choose the `TOCK_TARGETS` definition to be:
```
rv32i|rv32i.0x00080060.0x40008000|0x00080060|0x40008000
```

Thus, the first application's RAM region starts 0x4000 (16kb) above the kernel's `_sappmem`. If we were to set it to `0x40004000` (== `_sappmem`) even just enabling system call tracing adds an additional kB of kernel RAM and thus the application now expects it's RAM section in the middle of the kernel's RAM section which the process loader rightfully refuses.

My question was whether my observations above are correct and whether we have a "standard offset" between `_sappmem` of a vanilla kernel and the target configuration for RISC-V apps in `libtock-c`.

**Comment by bradjc** [comment]
Right, gotcha. We don't have any convention, I suspect we place the RAM address at wherever the kernel tells us it should be if you get it wrong (aka the first available address).

**Comment by lschuermann** [comment]
Alright, makes sense, that's in line with what I've seen in other boards. That being said, the LiteX boards have a high-degree of configurability given you're enabled and invited to change the actual hardware and present peripherals. Thus in tock/libtock-c#231 I've taken the liberty to add some padding to the kernel's `_sappmem` in the target definitions. These boards have -- by default -- a huge amount of RAM (256MB), thus this shouldn't be an issue.

Is there anything else blocking this PR, something I didn't think of?

**Comment by bradjc** [comment]
I decided it would be safer to have a default max_size, as I could see someone accidentally creating a really large file using this.

**Comment by lschuermann** [comment]
> I decided it would be safer to have a default max_size, as I could see someone accidentally creating a really large file using this.

I think that's a good idea. I don't even think we'd need to revisit this later on, as a board with a larger flash section could just define its own custom `max_size`, right? If I understand this correctly that should always override the default.

**Pull request merged**